### PR TITLE
Refactor controller unit test with proper mock

### PR DESCRIPTION
**Pull request by zacharya**
**Is this a bug fix or adding new feature?**
Refactor Cloud interface unit tests to use GoMock

**What is this PR about? / Why do we need it?**
- Addresses #220 .  
- One thing to discuss - moving `cloud.FakeCloudProvider` to `sanity.fakeCloudProvider` required exporting `cloud.metadata` due to the dependency.  I don't see any issues with this, but I figured I'd bring it up in case some of the maintainers see problems.

**What testing is done?** 
Both `make test` and `make test-sanity` are running fine.

### CODE REVIEW

**Comment by leakingtapan** [comment]
Im okay with making metadata struct public.

File: pkg/driver/controller_test.go [hunk 1]
```diff
@@ -48,327 +55,752 @@ func TestCreateVolume(t *testing.T) {
...
+					VolumeId: "invalid-volume-name",
+				}
+				expResp := &csi.DeleteVolumeResponse{}
+
+				ctx := context.Background()
+				mockCtl := gomock.NewController(t)
+				defer mockCtl.Finish()
+
+				mockCloud := mocks.NewMockCloud(mockCtl)
+				mockCloud.EXPECT().DeleteDisk(gomock.Eq(ctx), gomock.Eq(req.VolumeId)).Return(true, nil)
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: I think this case should return `cloud.ErrNotFound` for invalid volume ID.

File: pkg/driver/controller_test.go [hunk 2]
```diff
@@ -48,327 +55,752 @@ func TestCreateVolume(t *testing.T) {
...
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					t.Fatalf("Unexpected error: %v", srvErr.Code())
+				}
+				if !reflect.DeepEqual(resp, expResp) {
+					t.Fatalf("Expected resp to be %+v, got: %+v", expResp, resp)
+				}
 			},
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: And could you add a test case for `DeleteDisk` failed?

File: pkg/driver/controller_test.go [hunk 3]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+				if _, err := awsDriver.CreateSnapshot(context.Background(), req); err != nil {
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: Should add an else branch to fail the test if `err == nil`

File: pkg/driver/controller_test.go [hunk 4]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+
+				awsDriver := controllerService{cloud: mockCloud}
+				resp, err := awsDriver.CreateSnapshot(context.Background(), req)
+				if err != nil {
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+				}
+				if expErrCode != codes.OK {
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: This line is always true?
```suggestion
				if err != codes.OK {
```

File: pkg/driver/controller_test.go [hunk 5]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+				}
+
+				mockCloud.EXPECT().GetSnapshotByName(gomock.Eq(ctx), gomock.Eq(extraReq.GetName())).Return(mockSnapshot, nil)
+				_, err = awsDriver.CreateSnapshot(ctx, extraReq)
+				if err != nil {
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != extraExpErrCode {
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: inline the `extraExpErrCode` for readability since its actual codes.Ok

File: pkg/driver/controller_test.go [hunk 6]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+				}
+				snap := resp.GetSnapshot()
+				if snap == nil {
+					t.Fatalf("Expected snapshot %v, got nil", expSnapshot)
+				}
+
+				mockCloud.EXPECT().GetSnapshotByName(gomock.Eq(ctx), gomock.Eq(extraReq.GetName())).Return(mockSnapshot, nil)
+				_, err = awsDriver.CreateSnapshot(ctx, extraReq)
+				if err != nil {
+					srvErr, ok := status.FromError(err)
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: Could we failed right away if `err!=nil`?

File: pkg/driver/controller_test.go [hunk 7]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != extraExpErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
+				if extraExpErrCode != codes.OK {
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]:
```suggestion
				if err != codes.OK {
```

File: pkg/driver/controller_test.go [hunk 8]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
+				if expErrCode != codes.OK {
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]:
```suggestion
				if err != codes.OK {
```

File: pkg/driver/controller_test.go [hunk 9]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+		name     string
+		testFunc func(t *testing.T)
 	}{
 		{
-			name:       "success normal",
-			req:        &csi.DeleteSnapshotRequest{},
-			expErrCode: codes.OK,
+			name: "success normal",
+			testFunc: func(t *testing.T) {
+				expErrCode := codes.OK
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: NIT: inline `expErrCode`

File: pkg/driver/controller_test.go [hunk 10]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
+				if expErrCode != codes.OK {
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]:
```suggestion
				if err != codes.OK {
```

File: pkg/driver/controller_test.go [hunk 11]
```diff
@@ -432,166 +864,299 @@ func TestPickAvailabilityZone(t *testing.T) {
...
+					t.Fatalf("Expected error %v, got no error", expErrCode)
+				}
+			},
 		},
 		{
 			name: "success not found",
-			req: &csi.DeleteSnapshotRequest{
-				SnapshotId: "xxx",
+			testFunc: func(t *testing.T) {
+				expErrCode := codes.OK
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: NIT: inline `expErrCode`

File: pkg/driver/controller_test.go [hunk 12]
```diff
@@ -600,193 +1165,390 @@ func TestControllerPublishVolume(t *testing.T) {
...
+					srvErr, ok := status.FromError(err)
+					if !ok {
+						t.Fatalf("Could not get error status code from error: %v", srvErr)
+					}
+					if srvErr.Code() != expErrCode {
+						t.Fatalf("Expected error code %d, got %d message %s", expErrCode, srvErr.Code(), srvErr.Message())
+					}
+					return
+				}
+				if expErrCode != codes.OK {
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: This check seem problematic. Why do we need this check when `err == nil` from `if _, err := awsDriver.ControllerPublishVolume(ctx, req); `?

File: pkg/driver/controller_test.go [hunk 13]
```diff
@@ -600,193 +1165,390 @@ func TestControllerPublishVolume(t *testing.T) {
...
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: same above

File: pkg/driver/controller_test.go [hunk 14]
```diff
@@ -600,193 +1165,390 @@ func TestControllerPublishVolume(t *testing.T) {
...
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: same above

File: pkg/driver/controller_test.go [hunk 15]
```diff
@@ -600,193 +1165,390 @@ func TestControllerPublishVolume(t *testing.T) {
...
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]:
```suggestion
				if err == nil {
				   t.Fatalf("Expected error %v, got no error", expErrCode) 
				}
```

File: pkg/driver/controller_test.go [hunk 16]
```diff
@@ -600,193 +1165,390 @@ func TestControllerPublishVolume(t *testing.T) {
...
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: Same above

File: pkg/driver/controller_test.go [hunk 17]
```diff
@@ -600,193 +1165,390 @@ func TestControllerPublishVolume(t *testing.T) {
...
```
Comment (leakingtapan) [inline pkg/driver/controller_test.go]: same above

File: pkg/driver/node_test.go [hunk 18]
```diff
@@ -762,9 +808,9 @@ func TestNodeGetInfo(t *testing.T) {
 	}
 }
 
-func newTestNodeService(metadata cloud.MetadataService, mounter mount.Interface) nodeService {
+func newTestNodeService(cloud cloud.Cloud, mounter mount.Interface) nodeService {
```
Comment (leakingtapan) [inline pkg/driver/node_test.go]: I think we can just pass in Metadata here since it is changed to be public:

```suggestion
func newTestNodeService(cloud cloud.MetadataService, mounter mount.Interface) nodeService {
```

So that it saves us from creating mock cloud in the test as well.

**Comment by zacharya** [comment]
Thanks for the review.  I'll knock these out first thing tomorrow.

File: pkg/driver/node_test.go (hunk 18 above)
Comment (zacharya) [reply pkg/driver/node_test.go]: Good catch.  Done and was able to remove all mock cloud creation.

File: pkg/driver/controller_test.go (hunk 15 above)
Comment (zacharya) [reply pkg/driver/controller_test.go]: Done

File: pkg/driver/controller_test.go (hunk 16 above)
Comment (zacharya) [reply pkg/driver/controller_test.go]: Done

File: pkg/driver/controller_test.go (hunk 17 above)
Comment (zacharya) [reply pkg/driver/controller_test.go]: Done

File: pkg/driver/controller_test.go (hunk 14 above)
Comment (zacharya) [reply pkg/driver/controller_test.go]: Done

File: pkg/driver/controller_test.go (hunk 13 above)
Comment (zacharya) [reply pkg/driver/controller_test.go]: Done

File: pkg/driver/controller_test.go (hunk 12 above)
Comment (zacharya) [reply pkg/driver/controller_test.go]: I agree.  I think this one was just trying to port the dynamic error handling logic to the single test without enough context.  Removed.

File: pkg/driver/controller_test.go (hunk 11 above)
Comment (zacharya) [reply pkg/driver/controller_test.go]: Done

**Comment by zacharya** [comment]
@leakingtapan I think I've touched on all of your suggestions.  I went through each test case with the goal of making each more readable (rather than bring the dynamic error handling logic over from the previous state).  Let me know if you see anything that still needs fixing.

**Comment by leakingtapan** [comment]
The code a lot readable and more test covered now Thank @zacharya 

/lgtm
/approve

**Comment by bertinatto** [comment]
Just a heads up about this PR: we should always squash commits as much as possible.

**Comment by leakingtapan** [comment]
+1 Forgot to remind @zacharya of doing this.

**Comment by zacharya** [comment]
@bertinatto @leakingtapan Will do next time
